src/java/org/apache/cassandra/index/sai/disk/v2/V2VectorIndexSearcher.java

```diff
@@ -336,82 +336,73 @@ private SparseFixedBitSet bitSetForSearch()
+        // we want to build a bitset of ordinals corresponding to the rows.  We won't know which
+        // path to take until we have an accurate key count.
+        SparseFixedBitSet bits = bitSetForSearch();
+        IntArrayList rowIds = new IntArrayList();
+        try (var primaryKeyMap = primaryKeyMapFactory.newPerSSTablePrimaryKeyMap();
+             var ordinalsView = graph.getOrdinalsView())
         {
-            var maxSegmentRowId = metadata.toSegmentRowId(metadata.maxSSTableRowId);
+            for (PrimaryKey primaryKey : keysInRange)
+            {
+                long sstableRowId = primaryKeyMap.exactRowIdForPrimaryKey(primaryKey);
+                // skip rows that are not in our segment (or more preciesely, have no vectors that were indexed)
+                // or are not in this segment (exactRowIdForPrimaryKey returns a negative value for not found)
+                if (sstableRowId < metadata.minSSTableRowId)
+                    continue;
 
-            // if we are brute forcing, we want to build a list of segment row ids, but if not,
-            // we want to build a bitset of ordinals corresponding to the rows
-            SparseFixedBitSet bits = null;
-            IntArrayList rowIds = null;
-            if (numRows <= maxBruteForceRows)
-                rowIds = new IntArrayList();
-            else
-                bits = bitSetForSearch();
+                // if sstable row id has exceeded current ANN segment, stop
+                if (sstableRowId > metadata.maxSSTableRowId)
+                    break;
 
-            try (var ordinalsView = graph.getOrdinalsView())
-            {
-                for (PrimaryKey primaryKey : keysInRange)
-                {
-                    long sstableRowId = primaryKeyMap.exactRowIdForPrimaryKey(primaryKey);
-                    // skip rows that are not in our segment (or more preciesely, have no vectors that were indexed)
-                    // or are not in this segment (exactRowIdForPrimaryKey returns a negative value for not found)
-                    if (sstableRowId < metadata.minSSTableRowId)
-                        continue;
-
-                    // if sstable row id has exceeded current ANN segment, stop
-                    if (sstableRowId > metadata.maxSSTableRowId)
-                        break;
-
-                    int segmentRowId = metadata.toSegmentRowId(sstableRowId);
-                    if (numRows <= maxBruteForceRows)
-                    {
-                        rowIds.add(segmentRowId);
-                    }
-                    else
-                    {
-                        int ordinal = ordinalsView.getOrdinalForRowId(segmentRowId);
-                        if (ordinal >= 0)
-                            bits.set(ordinal);
-                    }
-                }
+                int segmentRowId = metadata.toSegmentRowId(sstableRowId);
+                rowIds.add(segmentRowId);
```

Comment:
Is it worth only adding to `rowIds` when `ordinal >= 0`? As my other comment/question indicates, I'm not sure if there is a case when we get a row id that won't have an ordinal.